### PR TITLE
Save memory if spark options are set accordingly

### PR DESCRIPTION
**Pull request by MarcMil**

### CODE REVIEW

**Comment by MarcMil** [comment]
These maps need a lot of memory. If someone needs them, they should subclass the PAG and manage the maps there.

**Comment by MarcMil** [comment]
OK, it seems someone uses these. I'll do the subclassing. Closing this for now. My IDE failed to show me the usage, otherwise I wouldn't have committed this early. Good thing we have these checks :D

File: src/main/java/soot/jimple/spark/pag/PAG.java [hunk 1]
```diff
@@ -1087,11 +1091,14 @@ public void addCallTarget(Edge e) {
...
       addEdge(parm, thiz);
       pval = addInterproceduralAssignment(parm, thiz, e);
-      callAssigns.put(ie, pval);
-      callToMethod.put(ie, srcmpag.getMethod());
 
-      if (virtualCall && !virtualCallsToReceivers.containsKey(ie)) {
-        virtualCallsToReceivers.put(ie, parm);
+      if (callAssigns != null) {
+        callToMethod.put(ie, srcmpag.getMethod());
+        boolean virtualCall = callAssigns.containsKey(ie);
```
Comment (StevenArzt) [inline src/main/java/soot/jimple/spark/pag/PAG.java:1097]: Wieso kein ``putIfAbsent()``?

File: src/main/java/soot/jimple/spark/pag/PAG.java [hunk 2]
```diff
@@ -1397,10 +1422,13 @@ public void addCallTarget(MethodPAG srcmpag, MethodPAG tgtmpag, Stmt s, Context
       thisRef = thisRef.getReplacement();
       addEdge(baseNode, thisRef);
       Pair<Node, Node> pval = addInterproceduralAssignment(baseNode, thisRef, e);
-      callAssigns.put(ie, pval);
-      callToMethod.put(ie, srcmpag.getMethod());
-      if (virtualCall && !virtualCallsToReceivers.containsKey(ie)) {
-        virtualCallsToReceivers.put(ie, baseNode);
+      if (callAssigns != null) {
+        boolean virtualCall = callAssigns.containsKey(ie);
```
Comment (StevenArzt) [inline src/main/java/soot/jimple/spark/pag/PAG.java:1426]: Wieso kein ``putIfAbsent()``?

**Review by StevenArzt — CHANGES_REQUESTED** [review]

File: src/main/java/soot/jimple/spark/pag/PAG.java (hunk 1 above)
Comment (MarcMil) [reply src/main/java/soot/jimple/spark/pag/PAG.java]: Es gibt kein putIfAbsent, weil es eine HashMultiMap ist. Aber man kann das tatsächlich mit einem put hinbekommen.

**Pull request merged**